### DISCUSSION TITLE
Separate modules

### DISCUSSION THREAD

**Discussion by aarondcohen** [aarondcohen/id128#9] [Ideas]
Probably less of an issue for server-side development, but there's a cost to adding new functionality for the client since the files size goes up.  Is it worth providing multiple smaller modules that just export a single id type?

Feedback from developers would be helpful.

**Comment by bjeanes** [comment]
True, but Webpack et al should be able to effectively tree-shake. Definitely an issue if directly being included though.

**Comment by RobertWHurst** [comment]
My recommendation would be to use https://lernajs.io/ to publish a package for each algorithm. Lerna will help you keep the packages in sync. There are is a significant number of popular projects using Lerna => https://lernajs.io/#users. Pug is a good example of Lerna's use => https://github.com/pugjs/pug

If this isn't something you wish to tackle, you could provide a file named after each algorithm at the root of your project folder. This would allow users to require just the algorithm they want by providing it as a subpath to your module. Example: `import Ulid from 'id128/ulid'` or `const Ulid = require('id128/ulid')`. This way Webpack and/or Parcel will not include unneeded files.

In any case, splitting up the algorithms would be awesome.

**Comment by aarondcohen** [comment]
I looked into lernajs and it looks pretty solid.  The biggest considerations to breaking apart the library are:
- within an id family, there's a lot of reuse in the form of inheritance
- some of the uuid functionality requires knowledge of the various algorithms to correctly parse a uuid

If you'd like to take a stab at this let me know.  I'm happy to discuss any issues that arise.
